process_report/invoices/billable_invoice.py

```diff
@@ -0,0 +1,186 @@
+from dataclasses import dataclass
+from decimal import Decimal
+import logging
+import sys
+
+import pandas
+import pyarrow
+
+import process_report.invoices.invoice as invoice
+import process_report.util as util
+
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+@dataclass
+class BillableInvoice(invoice.Invoice):
+    nonbillable_pis: list[str]
+    nonbillable_projects: list[str]
+    old_pi_filepath: str
+
+    def _prepare(self):
+        self.data = self._remove_nonbillables(
+            self.data, self.nonbillable_pis, self.nonbillable_projects
+        )
+        self.data = self._validate_pi_names(self.data)
+
+    def _process(self):
+        old_pi_df = self._load_old_pis(self.old_pi_filepath)
+        self.data, updated_old_pi_df = self._apply_credits_new_pi(self.data, old_pi_df)
+        self._dump_old_pis(self.old_pi_filepath, updated_old_pi_df)
+
+    def _remove_nonbillables(
+        self,
+        data: pandas.DataFrame,
+        nonbillable_pis: list[str],
+        nonbillable_projects: list[str],
```

Comment:
couldn't you access all these arguments with `self.data`, `self.nonbillable_pis`?

Reply:
I suppose I could access the class attributes directly. I copied most of these invoice functions from `process_report.py` without changing the signature since I thought:
1. There will be less changes needed for the unit tests
2. I envisioned that only `_prepare`, `_process`, and `_prepare_process` would be the only functions that would reference their object attributes, with any other class function behaving as static functions rather than class functions. I had a vague feeling that this is right, and will follow your opinion.